c7n/resources/vpc.py

```diff
@@ -2708,3 +2708,65 @@ class resource_type(query.TypeInfo):
+    will flagged.
+    This filter is useful for cost optimization and performance use-cases, where we don't want network traffic to cross
+    from one availability zone (AZ) to another AZ.
+
+    :Example:
+    .. code-block:: yaml
+            policies:
+              - name: cross-az-nat-gateway-traffic
+                resource: aws.route-table
+                filters:
+                    - type: cross-az-nat-gateway-route
+                actions:
+                  - notify
+    """
+    schema = type_schema('cross-az-nat-gateway-route')
+    permissions = ("ec2:DescribeRouteTables", "ec2:DescribeNatGateways", "ec2:DescribeSubnets")
+
+    def process(self, resources, event=None):
+        # dump of all subnets and nat-gateways to avoid multiple API calls
+        all_subnets = self.manager.get_resource_manager('aws.subnet').resources()
+        all_nat_gws = self.manager.get_resource_manager('nat-gateway').resources()
+        # Create subnet_id AZ mapping
+        subnets_az_map = {subnet["SubnetId"]: subnet["AvailabilityZone"] for subnet in all_subnets}
+        # Create nat_gateway subnet_id AZ mapping vpc wise
+        nat_gws_subnet_map = {nat_gateway['NatGatewayId']: nat_gateway["SubnetId"]
+                              for nat_gateway in all_nat_gws}
+        # List of AZ where NAT gateways are created vpc wise. To make sure NAT Gateway is available in a given AZ for a vpc
+        all_nat_gw_az = {}  # To store NAT Gateway subnet AZ per VPC ID
+        for nat_gateway in all_nat_gws:
+            vpc_id = nat_gateway["VpcId"]
+            subnet_az = subnets_az_map[nat_gateway["SubnetId"]]
+            if vpc_id in all_nat_gw_az:
+                all_nat_gw_az[vpc_id].append(subnet_az)
+            else:
+                all_nat_gw_az[vpc_id] = [subnet_az]
+        results = []  # To return filtered resources
+        for res in resources:
+            vpc_id = res["VpcId"]
+            res['NatGatewayAvailabilityZone'] = {}
+            for route in res["Routes"]:
+                if not route.get("NatGatewayId") or route.get("State") != "active":
+                    continue
+                nat_gw_az = subnets_az_map[nat_gws_subnet_map[route.get("NatGatewayId")]]
+                res['NatGatewayAvailabilityZone'][route.get("NatGatewayId")] = nat_gw_az
+                for association in res["Associations"]:
+                    subnet_id = association.get("SubnetId")
+                    if not subnet_id:
+                        continue
+                    association['SubnetAvailabilityZone'] = subnets_az_map[subnet_id]
+                    association['NatGatewayAvailableInSubnetAvailabilityZone'] = all_nat_gw_az.get(vpc_id) is not None and subnets_az_map[subnet_id] in all_nat_gw_az.get(vpc_id)
+                    # check if Associated Subnet AZ is same as NAT GW AZ
+                    if subnets_az_map[subnet_id] != nat_gw_az:
+                        results.append(res)
```

Comment:
A subnet can be associated with only one route-table at a time. Logic is to check if a NAT gateway is created in availability zone of the associated subnet. In case of multiple subnet association it will be set True if NAT gateway is created in AZ of any of the associated subnets.